### DISCUSSION TITLE
Custome route for update not working not working with latest livewire3

### DISCUSSION THREAD

**Discussion by aliworld1010** [livewire/livewire#7039] [7. Bugs (archived)]
### Livewire version

3.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

i am using livewire 3 + laravel project with th foldername testproj in "C:\xampp\htdocs"  

Project Path 
C:\xampp\htdocs\testproj 

I set my custom livewire path as per Installation livewire documentation :

Livewire::setScriptRoute(function ($handle) {
    return Route::get('/testproj/vendor/livewire/livewire/dist/livewire.js', $handle);
});

&

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/testproj/vendor/livewire/livewire/dist/update', $handle);
});

Results with php artisan serve ======works everything fine

Results with xampp =   http://localhost/testproj /public/user/profile
POST http://localhost/testproj /vendor/livewire/livewire/dist/update 404 (Not Found)  

so the update not working but the livewire,js is working ... but without update livewire is useless.

In simple words if i say .. someone install jetstream with livewire fresh project and try to publish in sub folder of xampp it will not work , even after setting custom route path

I need this project to be on subfolder of my root webfolder .. i can not publish at root folder as there are some other project is published there ... please help me i am strudgling to solve this for last 3 days and it is making me really think to use livewire or not... u guys have done amazing job i am sure u can help fixing this too

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by Serveronet** [comment]


**Reply by aliworld1010** [reply]
still same after 3.0.5
Error in console after click on update
livewire.js?id=f41737f6:531     POST http://localhost/testproj/vendor/livewire/livewire/dist/update 404 (Not Found)

In Web.php  i set the route.

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('testproj/vendor/livewire/livewire/dist/update', $handle);
    });

Tell me pls how to fix it .. i am seriously thinking to leave livewire as i am not able to solve it...

**Comment by aliworld1010** [comment]
[testproj.zip](https://github.com/livewire/livewire/files/12888184/testproj.zip)


I have added this as well but it did not work

Livewire::setScriptRoute(function ($handle) {
    return Route::get('/testproj/vendor/livewire/livewire/dist/livewire.js', $handle);
});

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/testproj/vendor/livewire/livewire/dist/update', [HandleRequestsHandleRequests::class, 'handleUpdate']);
});




For better understanding I have attached my project code with simple counter example and custom route so u can debug..

**Comment by amirfrd** [comment]
I want to run my project in subfolder too but i get 404 for livewire/update route too

**Comment by mohamed-slh-dev** [comment]
put this code in web.php

// :: LivewireServerDeployment in subRoute
if (env('APP_ENV') == 'production') {

    Livewire::setUpdateRoute(function ($handle) {
        return Route::post(env('LIVEWIRE_UPDATE_PATH'), $handle);
    });


    Livewire::setScriptRoute(function ($handle) {
        return Route::get(env('LIVEWIRE_JAVASCRIPT_PATH'), $handle);
    });

} // end if


and in .env change the APP_ENV = production and APP_DEBUG = true and paste this code

LIVEWIRE_UPDATE_PATH=/(your sub folder name)/public/livewire/update
LIVEWIRE_JAVASCRIPT_PATH=/(your sub folder name)/public/livewire/livewire.js

**Reply by joshhanley** [reply]
@mohamed-slh-dev you shouldn't use `env()` outside of config files. Instead I would create a config file and add two items for update and javascript path there and reference your ENV variables there.

Then in your route definitions, use the config value instead. That will let you define a default path too.

**Comment by Muhanned** [comment]
**_Put this code in web.php_**


```php
Livewire::setScriptRoute(function ($handle) {
    return Route::get('/SiteName/vendor/livewire/livewire/dist/livewire.js', $handle);
});

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/SiteName/livewire/update', $handle);
});
```

**Comment by syamsarosa** [comment]
I had this issue before in my work. The `setScriptRoute` and `setUpdateRoute` didn't work. 

The project has two Laravel apps: `public-app` and `admin-app`. The `public-app` is a Laravel Livewire project.

Since both apps have a domain, I add a `ServerName` for `public-app.com` and point the `DocumentRoot` to `/public-app/public/`. Below is the sample code of `public-app` Apache vhost.

```bash
<VirtualHost public-app:80>
     ServerName public-app.com
     DocumentRoot /var/www/html/public-app/public
     ...
</VirtualHost>
```

You can install Let's Encrypt cert later and Let's Encrypt will add a redirect from HTTP to HTTPS.

You can set up a domain using hosts for local development. In Ubuntu, the path is `/etc/hosts`. In Windows 11, I think it's in `C:\Windows\System32\drivers\etc\hosts`. You need administrator permission to edit it. Once you open the file, add the below code

```bash
127.0.0.1	public-app.test
```

**Comment by nuocgansoi** [comment]
The correct syntax is:
```php
Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/test/livewire/update', $handle)->middleware([
        'web',
        LivewireUpdateHtmlMinify::class,    // additional middleware options
    ]);
});
```
See the function:
```php
Livewire\Mechanisms\HandleRequests\HandleRequests::boot()
```

**Comment by latzkoo** [comment]
All up to the route naming. Just assign a custom name to your custom route:

```
Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/custom/livewire/update', $handle)->name('custom.livewire.update');
});
```

**Reply by igmltd** [reply]
thanks - custom named route worked - when i went through the code i could see why.

@joshhanley should this be in the docs?

**Reply by devlagret** [reply]
Yeah, it works for me too 
@Joshhanley, there is no error when I use the custom update route while following the docs at the beginning - it somehow randomly stop working, Am i missing something?; and it somehow takes me hours to find this comment and work

I think there is a problem with `Livewire\Mechanisms\HandleRequests` that has been called too early, before any other provider has been called?

*I dont realy know how laravel work so sorry in advance

**Reply by joshhanley** [reply]
@igmltd and @devlagret yeah this issue started with a changed in Laravel. I submitted a PR with a fix a couple of days ago and the fix has been merged into Livewire today and will be in the next v3 release https://github.com/livewire/livewire/pull/9699

**Comment by jkokotinis** [comment]
We have a similar issue. We have our laravel app behind a proxy with a path (e.g. https://myapp.com/apply). The proxy applies strip prefix, so the app sees https://myapp.com. For this reason in our AppServiceProvider we have

`
\URL::forceRootUrl(config('app.url'));
`

and we also set the env as:

```
APP_URL=https://myapp.com/apply
ASSET_URL=${APP_URL}
```

The produced urls in the html are:

```
<script src="https://myapp.com/apply/livewire-a644aedb/livewire.min.js?id=8f29b734" 
data-csrf="1aXHPhkRPTaeLcdOwvIkIRgWDnzNzzxyBVJL6gXv" 
data-module-url="https://myapp.com/apply/livewire-a644aedb" 
data-update-uri="https://myapp.com/apply/apply/livewire-a644aedb/update" 
data-navigate-once="true"></script>
```

Our issue is that (**only**) the update-uri has the path "/apply" 2 times.

Using the function Livewire::setUpdateRoute still produces the wrong path.

Laravel 12, Livewire v4.2.1

**Reply by Akulatraxas** [reply]
have you found a solution? i am at the same spot and nothing works to fix this.

**Reply by jkokotinis** [reply]
I didn’t have time to debug it further, so I changed the app domain to remove the path.

**Reply by jkokotinis** [reply]
solved it like this

```   
document.addEventListener('DOMContentLoaded', function() {
            const script = document.querySelector('script[data-update-uri]');
            if (script) {
                console.log("applying livewire hack");
                const uri = new URL(script.getAttribute('data-update-uri'));

                // Edit below to get your correct url path
                const segments = uri.pathname.split('/').filter(Boolean);

                for (let len = 1; len <= Math.floor(segments.length / 2); len++) {
                    const prefix = segments.slice(0, len).join('/');
                    const next = segments.slice(len, len * 2).join('/');
                    if (prefix === next) {
                        segments.splice(0, len); // remove the duplicate prefix
                        break;
                    }
                }
                
                uri.pathname = '/' + segments.join('/');
                //end edit 

               script.setAttribute('data-update-uri', uri.toString());
            }
        });
```
